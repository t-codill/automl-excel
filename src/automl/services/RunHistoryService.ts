import { RunHistoryAPIs, RunHistoryAPIsModels } from "@vienna/runhistory";

import { flatMap, mergeWith, reduce, sortBy } from "lodash";
import { BasicTypes } from "../common/BasicTypes";
import { IDictionary } from "../common/IDictionary";
import { isRunCompleted } from "../common/utils/run";
import { safeParseJson } from "../common/utils/safeParseJson";
import { IArtifact } from "./ArtifactService";
import { IServiceBaseProps } from "./ServiceBase";
import { ServiceBaseNonArm } from "./ServiceBaseNonArm";

const artifactLocationRegV1 = /^aml:\/\/artifactId\/([^/]+)\/([^/]+)\/outputs\/(.*)$/;
const artifactLocationRegV2 = /^aml:\/\/artifactId\/([^/]+)\/([^/]+)\/(.*)$/;
export type IRunDtoWithExperimentName = RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto & { experimentName: string };
export type RunMetricType = BasicTypes | BasicTypes[] | IDictionary<BasicTypes[]>;

function getMetricValue(metric: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto): RunMetricType {
    if (!metric.cells
        || !metric.schema
        || !metric.schema.properties
        || !metric.schema.properties[0]
        || !metric.schema.properties[0].name) {
        return undefined;
    }
    if (metric.schema.properties.length === 1) {
        const propertyName = metric.schema.properties[0].name;
        // scalar
        if (metric.cells.length === 1) {
            return metric.cells[0][propertyName];
        }
        // array
        else {
            return metric.cells.map((c) => c[propertyName]);
        }
    }
    else {
        // create value from property: {property1: [], property2:[]}
        const tableValue: IDictionary<BasicTypes[]> = reduce<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsMetricSchemaPropertyDto, {}>(
            metric.schema.properties, (result, value) => {
                if (value.name) {
                    result[value.name] = [];
                }
                return result;
            }, {});
        return reduce<RunMetricType, {}>(metric.cells, (result, value) => {
            // push cells value to result
            // cell: {property1: 13}, result: {property1: [1], property2:[5]}
            // return {property1: [1,13], property2:[5]}
            mergeWith(result, value, (r, v) => {
                return [...r, v];
            });
            return result;
        }, tableValue);
    }
}
export class RunHistoryService extends ServiceBaseNonArm<RunHistoryAPIs> {
    constructor(props: IServiceBaseProps) {
        super(props, RunHistoryAPIs, props.discoverUrls.history);
    }
    public getArtifactBackedMetricNames(metrics: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto[]): Array<[string, IArtifact]> {
        return reduce(metrics, (result, m) => {
            if (m.metricType) {
                let artifactLocationReg: RegExp;
                switch (m.metricType) {
                    case "azureml.v1.residuals":
                    case "azureml.v1.predictions":
                    case "azureml.v1.confusion_matrix":
                    case "azureml.v1.accuracy_table":
                        artifactLocationReg = artifactLocationRegV1;
                        break;
                    case "azureml.v2.residuals":
                    case "azureml.v2.predictions":
                    case "azureml.v2.confusion_matrix":
                    case "azureml.v2.accuracy_table":
                        artifactLocationReg = artifactLocationRegV2;
                        break;
                    default:
                        return result;
                }
                const match = artifactLocationReg.exec(m.dataLocation || "");
                if (match && m.name) {
                    result.push(
                        [
                            m.name,
                            {
                                origin: match[1],
                                container: match[2],
                                path: match[3]
                            }
                        ]);
                }
            }
            return result;
        }, [] as Array<[string, IArtifact]>);
    }
    public mergeMetrics(
        metrics: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto[],
        artifactBackedMetricNames: string[],
        artifacts: Array<string | undefined>): IDictionary<RunMetricType> {
        return reduce<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto, IDictionary<RunMetricType>>(metrics, (result, m) => {
            if (m.name) {
                const index = artifactBackedMetricNames.indexOf(m.name);
                result[m.name] = (index >= 0 && artifacts[index])
                    ? safeParseJson(artifacts[index])
                    : getMetricValue(m);
            }
            return result;
        }, {});

    }

    public async getExperiment(experimentName: string): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsExperimentDto | undefined> {

        return this.send(async (client, abortSignal) => {
            return client.getExperiment(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, experimentName, {
                abortSignal
            });
        });
    }

    public async getRun(runId: string, experimentName: string): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto | undefined> {
        return this.send(async (client, abortSignal) => {
            return client.getRunDetails(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, experimentName, runId, {
                abortSignal
            });
        });
    }

    public async getRunMetrics(runId: string, experimentName: string): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto[] | undefined> {
        return this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationtoken) => {
            return client.getRunMetrics(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, experimentName, {
                abortSignal,
                continuationtoken,
                filter: `RunId eq '${runId}'`
            });
        });
    }

    public async getChildRuns(runId: string, experimentName: string): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto[] | undefined> {
        const runs = await this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationtoken) => {
            return client.getChildRuns(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, experimentName, runId,
                {
                    abortSignal,
                    continuationtoken
                });
        });
        if (!runs) {
            return undefined;
        }
        return sortBy(runs, (run) => run.properties ? parseInt(run.properties.iteration, 10) : undefined);
    }

    public async getChildRunMetrics(
        childRuns: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto[],
        metricsNames: string[],
        experimentName: string): Promise<Array<IDictionary<RunMetricType>> | undefined> {

        if (childRuns.length === 0) {
            return [];
        }

        const completedRuns = childRuns
            .filter(isRunCompleted);
        if (completedRuns.length < 1) {
            return childRuns.map(() => this.mergeMetrics([], [], []));
        }

        const metrics = await this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationToken) => {
            return client.queryRunMetrics(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, experimentName, {
                abortSignal,
                queryParams: {
                    filter: `(${
                        completedRuns
                            .map((run) => `RunId eq '${run.runId}'`)
                            .join(" or ")
                        }) and (${
                        metricsNames.map((name) => `name eq '${name}'`)
                            .join(" or ")
                        })`,
                    continuationToken,
                    top: 1000
                }
            });
        });
        if (!metrics) { return undefined; }

        return childRuns.map((run) => this.mergeMetrics(metrics.filter((m) => m.runId === run.runId), [], []));
    }

    public async listExperiments(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsExperimentDto[] | undefined> {
        return this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationtoken) => {
            return client.getExperiments(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, {
                abortSignal,
                continuationtoken
            });
        });
    }

    // TODO: This is a temporary function only used by UX-Runner
    // This will be removed after we expose ODATA filter for getRunList
    public async getRuns(experiment: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsExperimentDto): Promise<IRunDtoWithExperimentName[] | undefined> {
        const allRuns = await this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationtoken) => {
            return client.getRuns(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, experiment.name || "",
                {
                    abortSignal,
                    continuationtoken,
                    filter: `Hidden eq false and parentRunId eq null and runType eq "automl"`,
                    top: 1000
                });
        });
        if (!allRuns) {
            return undefined;
        }

        return allRuns.map(
            ((run) => (
                {
                    ...run,
                    experimentName: experiment.name || ""
                })));
    }

    public async getRunList(): Promise<IRunDtoWithExperimentName[] | undefined> {
        const allExperiments = await this.listExperiments();
        if (!allExperiments) {
            return undefined;
        }
        const allRuns = await this.parallelGetAllValuesWithContinuationToken(allExperiments, async (exp, client, abortSignal, continuationtoken) => {
            return client.getRuns(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, exp.name || "",
                {
                    abortSignal,
                    continuationtoken,
                    filter: `Hidden eq false and parentRunId eq null and runType eq "automl"`,
                    top: 1000
                });
        });
        if (!allRuns) {
            return undefined;
        }

        return flatMap(
            allRuns.map(
                (com, idx) =>
                    com.map((run) => (
                        {
                            ...run,
                            experimentName: allExperiments[idx].name || ""
                        }))));
    }

    public async updateTag(
        run: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto,
        experimentName: string,
        name: string,
        value: string
    ): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto | undefined> {
        return this.send(async (client, abortSignal) => {
            if (!run.runId) {
                return undefined;
            }
            const tags = { ...run.tags };
            tags[name] = value;
            return client.addOrModifyRun(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                experimentName,
                run.runId,
                {
                    createRunDto: {
                        runId: run.runId,
                        tags
                    },
                    abortSignal
                });
        });
    }

    public async updateTags(
        run: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto,
        experimentName: string,
        names: string[],
        values: string[]
    ): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto | undefined> {
        if (!names || !values || names.length !== values.length) {
            return undefined;
        }
        return this.send(async (client, abortSignal) => {
            if (!run.runId) {
                return undefined;
            }
            const tags = { ...run.tags };
            for (let i = 0; i < names.length; i++) {
                tags[names[i]] = values[i];
            }
            return client.addOrModifyRun(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                experimentName,
                run.runId,
                {
                    createRunDto: {
                        runId: run.runId,
                        tags
                    },
                    abortSignal
                });
        });
    }
}
