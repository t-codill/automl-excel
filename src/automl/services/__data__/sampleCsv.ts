export const sampleCsv = `a,b,c,d,e
1,2,3,4,5
1,2,3,4,5
1,2,3,4,5
1,2,3,4,5
1,2,3,4,5
`;

const largeCsvHeader = Array(50)
    .fill("aaaaa,bbbbb,ccccc,ddddd,eeeee")
    .join(",");

const largeCsvRow = Array(50)
    .fill("11111,22222,33333,44444,55555")
    .join(",");

export const largeCsv = `${largeCsvHeader}
${Array(50)
        .fill(largeCsvRow)
        .join("\n")}
`;
