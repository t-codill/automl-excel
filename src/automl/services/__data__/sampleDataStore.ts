import { AzureMachineLearningDatastoreManagementModels } from "@vienna/datastore";
import { NotNullableProperties } from "../../common/utils/NotNullableProperties";

export const sampleDataStore: NotNullableProperties<AzureMachineLearningDatastoreManagementModels.DataStoreDto> = {
    dataStoreType: "AzureBlob",
    name: "defaultDataStore",
    azureStorageSection: {
        accountKey: "key",
        accountName: "account",
        credentialType: "AccountKey",
        credential: "secret",
        containerName: "defaultContainer",
        isSas: false
    },
    azureDataLakeSection: {},
    azurePostgreSqlSection: {},
    azureSqlDatabaseSection: {},
    hasBeenValidated: true
};
