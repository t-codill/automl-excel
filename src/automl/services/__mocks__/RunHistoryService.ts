import { RunHistoryAPIsModels } from "@vienna/runhistory";
import { classificationSuccessRun } from "../../childRun/__data__/classificationSuccessRun";
import { BasicTypes } from "../../common/BasicTypes";
import { IDictionary } from "../../common/IDictionary";
import { parentSuccessRun } from "../../parentRun/__data__/parentSuccessRun";
import { IArtifact } from "../ArtifactService";

export type IRunDtoWithExperimentName = RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto & { experimentName: string };
export type RunMetricType = BasicTypes | BasicTypes[] | IDictionary<BasicTypes[]>;

export class RunHistoryService {
    public getArtifactBackedMetricNames(): Array<[string, IArtifact]> {
        const artifact: IArtifact = {
            origin: "",
            container: "",
            path: ""
        };
        return [["artifact", artifact]];
    }
    public mergeMetrics(): IDictionary<RunMetricType> {
        if (classificationSuccessRun.runMetrics) {
            return classificationSuccessRun.runMetrics;
        }
        return {};
    }

    public async getExperiment(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsExperimentDto | undefined> {
        return undefined;
    }

    public async getRun(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto | undefined> {
        return classificationSuccessRun.run;
    }

    public async getRunMetrics(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto[] | undefined> {
        return [classificationSuccessRun.runMetrics];
    }

    public async getChildRuns(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto[] | undefined> {
        return undefined;
    }

    public async getChildRunMetrics(): Promise<Array<IDictionary<RunMetricType>> | undefined | undefined> {
        return undefined;
    }

    public async listExperiments(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsExperimentDto[] | undefined> {
        return [
            {
                createdUtc: new Date(2019, 1, 1, 1, 1, 1, 1),
                description: "Description 1",
                experimentId: "Experiment Id 1",
                name: "Experiment Name 1"
            },
            {
                createdUtc: new Date(2019, 2, 2, 2, 2, 2, 2),
                description: "Description 2",
                experimentId: "Experiment Id 2",
                name: "Experiment Name 2"
            },
            {}
        ];
    }

    public async getRunList(): Promise<IRunDtoWithExperimentName[] | undefined> {
        return [
            {
                experimentName: "experiment1",
                runId: "run1"
            },
            {
                experimentName: "experiment2",
                runId: "run2"
            }
        ];
    }

    public async updateTag(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto | undefined> {
        return parentSuccessRun.run;
    }

    public async updateTags(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto | undefined> {
        return parentSuccessRun.run;
    }

    public dispose(): void {
        return;
    }

    public reset(): void {
        return;
    }
}
