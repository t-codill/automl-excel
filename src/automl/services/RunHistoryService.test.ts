import { RunHistoryAPIs } from "@vienna/runhistory";
import { restCanceledError } from "../../__data__/restCanceledError";
import { testContext } from "../common/context/__data__/testContext";
import { parentSuccessRun } from "../parentRun/__data__/parentSuccessRun";
import { RunHistoryService } from "./RunHistoryService";

jest.mock("@vienna/runhistory");

const service = new RunHistoryService(testContext);

describe("RunHistoryService", () => {
    it("should getExperiment", async () => {
        const result = await service.getExperiment("sampleExperimentName");
        expect(result)
            .toMatchSnapshot();
    });
    it("should getRun", async () => {
        const result = await service.getRun("00000000-0000-0000-0000-000000000000", "sampleExperimentName");
        expect(result)
            .toMatchSnapshot();
    });
    it("should getRunMetrics", async () => {
        const result = await service.getRunMetrics("00000000-0000-0000-0000-000000000000", "sampleExperimentName");
        expect(result)
            .toMatchSnapshot();
    });
    it("should getChildRuns", async () => {
        const result = await service.getChildRuns("00000000-0000-0000-0000-000000000000", "sampleExperimentName");
        expect(result)
            .toMatchSnapshot();
    });
    it("getChildRuns should sort by iteration", async () => {
        jest.spyOn(RunHistoryAPIs.prototype, "getChildRuns")
            .mockImplementationOnce(async () => ({
                value: [
                    {
                        runId: "AutoML_001",
                        properties: { iteration: "23" }
                    },
                    {
                        runId: "AutoML_000",
                        properties: { iteration: "9" }
                    }
                ]
            }));
        const result = await service.getChildRuns("00000000-0000-0000-0000-000000000000", "sampleExperimentName");
        expect(result)
            .toEqual([
                {
                    runId: "AutoML_000",
                    properties: { iteration: "9" }
                },
                {
                    runId: "AutoML_001",
                    properties: { iteration: "23" }
                }
            ]
            );
    });
    it("getChildRuns should return undefined if canceled", async () => {
        jest.spyOn(RunHistoryAPIs.prototype, "getChildRuns")
            .mockImplementationOnce(async () => { throw restCanceledError; });
        const result = await service.getChildRuns("00000000-0000-0000-0000-000000000000", "sampleExperimentName");
        expect(result)
            .toBeUndefined();
    });
    it("should listExperiments", async () => {
        const result = await service.listExperiments();
        expect(result)
            .toMatchSnapshot();
    });
    it("should update tag", async () => {
        const result = await service.updateTag(parentSuccessRun.run,
            parentSuccessRun.experimentName, "test tag", "test value");
        expect(result)
            .toMatchSnapshot();
    });
    it("should not update tag without run id", async () => {
        const result = await service.updateTag({ ...parentSuccessRun.run, runId: undefined },
            parentSuccessRun.experimentName, "test tag", "test value");
        expect(result)
            .toBeUndefined();
    });
    it("should update tags", async () => {
        const result = await service.updateTags(parentSuccessRun.run,
            parentSuccessRun.experimentName, ["test tag1", "test tag2", "test tag3"], ["test value1", "test value2", "test value3"]);
        expect(result)
            .toMatchSnapshot();
    });
    it("should not update tags without run id", async () => {
        const result = await service.updateTags({ ...parentSuccessRun.run, runId: undefined },
            parentSuccessRun.experimentName, ["test tag1", "test tag2", "test tag3"], ["test value1", "test value2", "test value3"]);
        expect(result)
            .toBeUndefined();
    });
    it("should not update tag without invalid key-value pair", async () => {
        const result = await service.updateTags(parentSuccessRun.run,
            parentSuccessRun.experimentName, ["test tag1", "test tag2", "test tag3"], ["test value1", "test value2"]);
        expect(result)
            .toMatchSnapshot();
    });
    describe("getChildRunMetrics", () => {
        it("should be empty without child run", async () => {
            const result = await service.getChildRunMetrics([], [], "sampleExperimentName");
            expect(result)
                .toEqual([]);
        });

        it("should be empty with no child run completed", async () => {
            const result = await service.getChildRunMetrics([
                { runId: "runId1", status: "Running" },
                { runId: "runId1", status: "Canceled" },
                { runId: "runId1", status: "Failed" }
            ], [], "sampleExperimentName");
            expect(result)
                .toEqual([{}, {}, {}]);
        });
        it("should be undefined when canceled", async () => {
            const spy = jest.spyOn(RunHistoryAPIs.prototype, "queryRunMetrics");
            spy.mockImplementationOnce(() => { throw restCanceledError; });
            const result = await service.getChildRunMetrics([{ runId: "runId1", status: "Completed" }], [], "sampleExperimentName");
            expect(result)
                .toBeUndefined();
        });
        it("should getChildRunMetrics", async () => {
            const result = await service.getChildRunMetrics([{ runId: "runId1", status: "Completed" }, { runId: "runId2", status: "Canceled" }], ["metric1"], "sampleExperimentName");
            expect(result)
                .toMatchSnapshot();
        });
    });

    describe("getArtifactBackedMetricNames", () => {
        it("should not get value without data location", async () => {
            const result = service.getArtifactBackedMetricNames([
                {
                    name: "noDataLocation",
                    metricType: "azureml.v1.residuals",
                    dataLocation: undefined
                }
            ]);
            expect(result)
                .toEqual([]);
        });
        it("should not get value with wrong location", async () => {
            const result = service.getArtifactBackedMetricNames([
                {
                    name: "metricName1",
                    dataLocation: "inlineMetric"
                }
            ]);
            expect(result)
                .toEqual([]);
        });
        it("should not value with unsupported metric type", async () => {
            const result = service.getArtifactBackedMetricNames([
                {
                    name: "metricName2",
                    metricType: "azureml.v1.image",
                    dataLocation: "aml://artifactId/ExperimentRun/AutoML_00000000-0000-0000-0000-000000000000_00/outputs/residuals"
                }
            ]);
            expect(result)
                .toEqual([]);
        });
        it("should get value with aml data location v1", async () => {
            const result = service.getArtifactBackedMetricNames([
                {
                    name: "metricName2",
                    metricType: "azureml.v1.residuals",
                    dataLocation: "aml://artifactId/ExperimentRun/AutoML_00000000-0000-0000-0000-000000000000_00/outputs/residuals"
                }
            ]);
            expect(result)
                .toEqual([
                    [
                        "metricName2",
                        {
                            container: "AutoML_00000000-0000-0000-0000-000000000000_00",
                            origin: "ExperimentRun",
                            path: "residuals",
                        },
                    ],
                ]);
        });
        it("should get value with aml data location v2", async () => {
            const result = service.getArtifactBackedMetricNames([
                {
                    name: "metricName2",
                    metricType: "azureml.v2.residuals",
                    dataLocation: "aml://artifactId/ExperimentRun/AutoML_00000000-0000-0000-0000-000000000000_00/residuals"
                }
            ]);
            expect(result)
                .toEqual([
                    [
                        "metricName2",
                        {
                            container: "AutoML_00000000-0000-0000-0000-000000000000_00",
                            origin: "ExperimentRun",
                            path: "residuals",
                        },
                    ],
                ]);
        });
    });
    describe("getRuns", () => {
        it("should getRunList", async () => {
            const spy = jest.spyOn(RunHistoryAPIs.prototype, "getRuns");
            const result = await service.getRuns({});
            expect(spy)
                .toBeCalledTimes(1);
            expect(result)
                .toMatchSnapshot();
        });
        it("should be undefined if canceled", async () => {
            const spy = jest.spyOn(RunHistoryAPIs.prototype, "getRuns");
            spy.mockImplementation(() => { throw restCanceledError; });
            const result = await service.getRuns({});
            expect(spy)
                .toBeCalledTimes(1);
            expect(result)
                .toBeUndefined();
        });
    });
    describe("getRunList", () => {
        it("should getRunList", async () => {
            const result = await service.getRunList();
            expect(result)
                .toMatchSnapshot();
        });
        it("should be undefined without experiment", async () => {
            const spy = jest.spyOn(RunHistoryAPIs.prototype, "getExperiments");
            spy.mockImplementation(() => { throw restCanceledError; });
            const result = await service.getRunList();
            expect(result)
                .toBeUndefined();
        });
        it("should be undefined if canceled", async () => {
            const spy = jest.spyOn(RunHistoryAPIs.prototype, "getRuns");
            spy.mockImplementation(() => { throw restCanceledError; });
            const result = await service.getRunList();
            expect(result)
                .toBeUndefined();
        });
    });
    describe("mergeMetrics", () => {

        describe("scalar value", () => {
            it("should not get metric without name", async () => {
                const result = service.mergeMetrics([
                    {
                        name: undefined,
                        dataLocation: "inlineMetric"
                    }
                ], [], []);
                expect(result)
                    .toEqual({});
            });
            it("should get undefined for metric without value", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "noValueMetric",
                        dataLocation: "inlineMetric"
                    },
                ], [], []);
                expect(result)
                    .toEqual({ noValueMetric: undefined });
            });
            it("should get for metric without data location", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "noLocationMetric",
                        dataLocation: undefined
                    },
                ], [], []);
                expect(result)
                    .toEqual({ noLocationMetric: undefined });
            });
            it("should get scalar metric", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "scalarMetric",
                        dataLocation: "inlineMetric",
                        schema: {
                            numProperties: 1,
                            properties: [{
                                name: "scalar",
                                propertyId: "scalar",
                                type: "scalar"
                            }]
                        },
                        cells: [{ scalar: "scalarValue" }]
                    },
                ], [], []);
                expect(result)
                    .toEqual({ scalarMetric: "scalarValue" });
            });
            it("should get undefined for scalar metric without property name", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "scalarMetricNoName",
                        dataLocation: "inlineMetric",
                        schema: {
                            numProperties: 1,
                            properties: [{
                                propertyId: "scalarMetricNoName",
                                type: "scalar"
                            }]
                        },
                        cells: [{ "": "scalarValue" }]
                    },
                ], [], []);
                expect(result)
                    .toEqual({ scalarMetricNoName: undefined });
            });
        });

        describe("array value", () => {
            it("should get array metric", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "arrayMetric",
                        dataLocation: "inlineMetric",
                        schema: {
                            numProperties: 1,
                            properties: [{
                                name: "array",
                                propertyId: "array",
                                type: "array"
                            }]
                        },
                        cells: [{ array: "arrayValue1" }, { array: "arrayValue2" }]
                    },
                ], [], []);
                expect(result)
                    .toEqual({ arrayMetric: ["arrayValue1", "arrayValue2"] });
            });
            it("should get undefined for array metric without schema", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "arrayMetricNoSchema",
                        dataLocation: "inlineMetric",
                        cells: [{ array: "arrayValue1" }, { array: "arrayValue2" }]
                    },
                ], [], []);
                expect(result)
                    .toEqual({ arrayMetricNoSchema: undefined });
            });
            it("should get undefined for array metric without schema properties", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "arrayMetricNoSchemaProperties",
                        dataLocation: "inlineMetric",
                        schema: {
                            numProperties: 0
                        },
                        cells: [{ array: "arrayValue1" }, { array: "arrayValue2" }]
                    },
                ], [], []);
                expect(result)
                    .toEqual({ arrayMetricNoSchemaProperties: undefined });
            });
            it("should get undefined for array metric without property 0", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "arrayMetricNoSchemaProperties0",
                        dataLocation: "inlineMetric",
                        schema: {
                            numProperties: 0,
                            properties: []
                        },
                        cells: [{ array: "arrayValue1" }, { array: "arrayValue2" }]
                    },
                ], [], []);
                expect(result)
                    .toEqual({ arrayMetricNoSchemaProperties0: undefined });
            });
            it("should get undefined for array metric without schema", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "arrayMetricNoSchemaProperties0Name",
                        dataLocation: "inlineMetric",
                        schema: {
                            numProperties: 1,
                            properties: [{
                                propertyId: "array",
                                type: "array"
                            }]
                        },
                        cells: [{ array: "arrayValue1" }, { array: "arrayValue2" }]
                    },
                ], [], []);
                expect(result)
                    .toEqual({ arrayMetricNoSchemaProperties0Name: undefined });
            });
        });
        describe("table value", () => {
            it("should get table value", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "tableMetric",
                        dataLocation: "inlineMetric",
                        schema: {
                            numProperties: 1,
                            properties: [
                                {
                                    name: "prop1",
                                    propertyId: "prop1",
                                    type: "table"
                                },
                                {
                                    name: "prop2",
                                    propertyId: "prop2",
                                    type: "table"
                                },
                                {
                                    propertyId: "noNameProp",
                                    type: "table"
                                }
                            ]
                        },
                        cells: [
                            { prop1: "scalarValue11" },
                            { prop1: "scalarValue12" },
                            { prop1: "scalarValue13" },
                            { prop1: "scalarValue14" },
                            { prop2: "scalarValue21" },
                            { prop2: "scalarValue22" },
                            { prop2: "scalarValue23" },
                            { prop2: "scalarValue24" }
                        ]
                    },
                ], [], []);
                expect(result)
                    .toEqual({
                        tableMetric: {
                            prop1: [
                                "scalarValue11",
                                "scalarValue12",
                                "scalarValue13",
                                "scalarValue14",
                            ],
                            prop2: [
                                "scalarValue21",
                                "scalarValue22",
                                "scalarValue23",
                                "scalarValue24",
                            ],
                        }
                    });
            });
        });

        describe("artifact value", () => {
            it("should get empty object for invalid json value", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "invalidArtifactMetric",
                        dataLocation: "aml://artifactId/ExperimentRun/AutoML_00000000-0000-0000-0000-000000000000_00/outputs/metric"
                    }
                ], ["invalidArtifactMetric"], ["invalid json"]);
                expect(result)
                    .toEqual({ invalidArtifactMetric: {} });
            });
            it("should get valid artifact metric", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "validArtifactMetric",
                        dataLocation: "aml://artifactId/ExperimentRun/AutoML_00000000-0000-0000-0000-000000000000_00/outputs/metric"
                    }
                ], ["validArtifactMetric"], [JSON.stringify({ value: 123 })]);
                expect(result)
                    .toEqual({ validArtifactMetric: { value: 123 } });
            });
            it("should get undefined if artifact does not exist", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "nonExistArtifactMetric",
                        dataLocation: "aml://artifactId/ExperimentRun/AutoML_00000000-0000-0000-0000-000000000000_00/outputs/metric"
                    }
                ], ["nonExistArtifactMetric"], []);
                expect(result)
                    .toEqual({ nonExistArtifactMetric: undefined });
            });
            it("should get undefined if artifact not found", async () => {
                const result = service.mergeMetrics([
                    {
                        name: "notFoundArtifactMetric",
                        dataLocation: "aml://artifactId/ExperimentRun/AutoML_00000000-0000-0000-0000-000000000000_00/outputs/metric"
                    }
                ], [], []);
                expect(result)
                    .toEqual({ nonExistArtifactMetric: undefined });
            });
        });
    });
});
