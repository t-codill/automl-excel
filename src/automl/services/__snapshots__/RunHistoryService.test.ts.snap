// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunHistoryService getChildRunMetrics should getChildRunMetrics 1`] = `
Array [
  Object {},
  Object {},
]
`;

exports[`RunHistoryService getRunList should getRunList 1`] = `
Array [
  Object {
    "experimentName": "experiment1",
    "runId": "AutoML_000",
  },
  Object {
    "experimentName": "",
    "runId": "AutoML_000",
  },
]
`;

exports[`RunHistoryService getRuns should getRunList 1`] = `
Array [
  Object {
    "experimentName": "",
    "runId": "AutoML_000",
  },
]
`;

exports[`RunHistoryService should getChildRuns 1`] = `
Array [
  Object {
    "runId": "AutoML_000",
  },
]
`;

exports[`RunHistoryService should getExperiment 1`] = `
Object {
  "_mockParameters": Array [
    "00000000-0000-0000-0000-000000000000",
    "testResource",
    "testWorkSpace",
    "sampleExperimentName",
    Object {
      "abortSignal": Object {},
    },
  ],
  "experimentId": "experimentId",
}
`;

exports[`RunHistoryService should getRun 1`] = `
Object {
  "_mockParameters": Array [
    "00000000-0000-0000-0000-000000000000",
    "testResource",
    "testWorkSpace",
    "sampleExperimentName",
    "00000000-0000-0000-0000-000000000000",
    Object {
      "abortSignal": Object {},
    },
  ],
  "runId": "AutoML_000",
}
`;

exports[`RunHistoryService should getRunMetrics 1`] = `
Array [
  Object {
    "name": "metric1",
  },
]
`;

exports[`RunHistoryService should listExperiments 1`] = `
Array [
  Object {
    "name": "experiment1",
  },
  Object {
    "name": undefined,
  },
]
`;

exports[`RunHistoryService should not update tag without invalid key-value pair 1`] = `undefined`;

exports[`RunHistoryService should update tag 1`] = `
Object {
  "_mockParameters": Array [
    "00000000-0000-0000-0000-000000000000",
    "testResource",
    "testWorkSpace",
    "Test",
    "AutoML_000",
    Object {
      "abortSignal": Object {},
      "createRunDto": Object {
        "runId": "AutoML_000",
        "tags": Object {
          "iterations": "2",
          "test tag": "test value",
        },
      },
    },
  ],
  "runId": "AutoML_000",
}
`;

exports[`RunHistoryService should update tags 1`] = `
Object {
  "_mockParameters": Array [
    "00000000-0000-0000-0000-000000000000",
    "testResource",
    "testWorkSpace",
    "Test",
    "AutoML_000",
    Object {
      "abortSignal": Object {},
      "createRunDto": Object {
        "runId": "AutoML_000",
        "tags": Object {
          "iterations": "2",
          "test tag1": "test value1",
          "test tag2": "test value2",
          "test tag3": "test value3",
        },
      },
    },
  ],
  "runId": "AutoML_000",
}
`;
